File: 2002-maximum-product-of-the-length-of-two-palindromic-subsequences/2002-maximum-product-of-the-length-of-two-palindromic-subsequences.py
class Solution:
    def maxProduct(self, s: str) -> int:
        res = 0

        def isPalindrome(sub):
            left, right = 0, len(sub) - 1

            while left < right:
                if sub[left] != sub[right]:
                    return False
                left += 1
                right -= 1
            
            return True
        
        def backtracking(sub1, sub2, idx):
            nonlocal res
            if idx == len(s):
                if isPalindrome(sub1) and isPalindrome(sub2):
                    res = max(res, len(sub1) * len(sub2))

                return

            backtracking(sub1 + s[idx], sub2, idx + 1)
            backtracking(sub1, sub2 + s[idx], idx + 1)
            backtracking(sub1, sub2, idx + 1)
        
        backtracking("", "", 0)

        return res

File: 2002-maximum-product-of-the-length-of-two-palindromic-subsequences/2002-maximum-product-of-the-length-of-two-palindromic-subsequences.java
class Solution {
    int maxProduct = 0;
    public int maxProduct(String s) {
        backtrack(s, 0, "", "");
        return maxProduct;
    }

    private void backtrack(String s, int idx, String sub1, String sub2) {
        if (idx == s.length()) {
            if (isPalindrom(sub1) && isPalindrom(sub2)) {
                maxProduct = Math.max(maxProduct, sub1.length() * sub2.length());
            }   
            return;
        }

        backtrack(s, idx + 1, sub1 + s.charAt(idx), sub2);
        backtrack(s, idx + 1, sub1, sub2 + s.charAt(idx));
        backtrack(s, idx + 1, sub1, sub2);
    }





    private boolean isPalindrom(String s) {
        int i = 0;
        int j = s.length() - 1;

        while (i < j) {
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}